packages/react-server/core/renderMiddleware.js

```diff
@@ -105,16 +105,25 @@ module.exports = function(server, routes) {
 			navigateDfd.resolve();
 
 			if (err) {
-				logger.log("onNavigate received a non-2xx HTTP code", err);
-				handleResponseComplete(req, res, context, start, page);
-				if (err.status && err.status === 404) {
-					next();
-				} else if (err.status === 301 || err.status === 302 || err.status === 307) {
+				// The page can elect to proceed to render
+				// even with a non-2xx response.  If it
+				// _doesn't_ do so then we're done.
+				var done = !(page && page.getHaveDocument());
+
+				if (err.status === 301 || err.status === 302 || err.status === 307) {
 					res.redirect(err.status, err.redirectUrl);
-				} else {
-					next(err);
+				} else if (done) {
```

Comment:
>  I think `res.redirect` adds the appropriate body

Oh, I see.  Thanks, I didn't realize that.  I'll push a patch to make `hasDocument` work for redirects.

> I don't think there's ever a case where we'd want to fall through

The case is: I want to customize the content of my redirect body.  It's cool that express provides a helper that populates a default, but I want the ability to render my own.

That means I'm responsible for actually putting the link in, but presumably if I'm claiming `hasDocument` on a redirect I know what I'm doing?